<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working Locally using anaconda, please uncomment the following code and execute it.
Use the version as per your python version.


In [1]:
!pip install yfinance
!pip install bs4
!pip install nbformat
!pip install --upgrade plotly

In [2]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [3]:
import plotly.io as pio
pio.renderers.default = "iframe"

In Python, you can ignore warnings using the warnings module. You can use the filterwarnings function to filter or ignore specific warning messages or categories.


In [4]:
import warnings
# Ignore all warnings
warnings.filterwarnings("ignore", category=FutureWarning)

## Define Graphing Function


In this section, we define the function `make_graph`. **You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.**


In [5]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021-06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()
    from IPython.display import display, HTML
    fig_html = fig.to_html()
    display(HTML(fig_html))

Use the make_graph function that we’ve already defined. You’ll need to invoke it in questions 5 and 6 to display the graphs and create the dashboard. 
> **Note: You don’t need to redefine the function for plotting graphs anywhere else in this notebook; just use the existing function.**


## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [6]:
import yfinance as yf

# Créer l'objet Ticker pour Tesla
tesla = yf.Ticker("TSLA")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to ` "max" ` so we get information for the maximum amount of time.


In [7]:
# Extraire l'historique des cours de Tesla
tesla_data = tesla.history(period="max")

# Réinitialiser l'index pour que la colonne Date devienne normale
tesla_data.reset_index(inplace=True)

# Afficher les premières lignes pour vérifier
tesla_data.head()

Date      Open      High       Low     Close  \
0 2010-06-29 00:00:00-04:00  1.266667  1.666667  1.169333  1.592667   
1 2010-06-30 00:00:00-04:00  1.719333  2.028000  1.553333  1.588667   
2 2010-07-01 00:00:00-04:00  1.666667  1.728000  1.351333  1.464000   
3 2010-07-02 00:00:00-04:00  1.533333  1.540000  1.247333  1.280000   
4 2010-07-06 00:00:00-04:00  1.333333  1.333333  1.055333  1.074000   

      Volume  Dividends  Stock Splits  
0  281494500        0.0           0.0  
1  257806500        0.0           0.0  
2  123282000        0.0           0.0  
3   77097000        0.0           0.0  
4  103003500        0.0           0.0

**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [8]:
import yfinance as yf
import pandas as pd

# Créer l'objet Ticker pour Tesla
tesla = yf.Ticker("TSLA")

# Extraire l'historique complet
tesla_data = tesla.history(period="max")

# Réinitialiser l'index pour que la colonne Date devienne normale
tesla_data.reset_index(inplace=True)

# Afficher les 5 premières lignes
tesla_data.tail()

Date        Open        High         Low  \
3828 2025-09-17 00:00:00-04:00  415.750000  428.309998  409.670013   
3829 2025-09-18 00:00:00-04:00  428.869995  432.220001  416.559998   
3830 2025-09-19 00:00:00-04:00  421.820007  429.470001  421.720001   
3831 2025-09-22 00:00:00-04:00  431.109985  444.980011  429.130005   
3832 2025-09-23 00:00:00-04:00  439.880005  440.970001  423.720001   

           Close     Volume  Dividends  Stock Splits  
3828  425.859985  106133500        0.0           0.0  
3829  416.850006   90454500        0.0           0.0  
3830  426.070007   93131000        0.0           0.0  
3831  434.209991   97108800        0.0           0.0  
3832  425.850006   83211500        0.0           0.0

## Question 2: Use Webscraping to Extract Tesla Revenue Data


In [ ]:
import requests

# URL de la page à télécharger
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"

# Envoyer la requête GET
response = requests.get(url)

# Enregistrer le texte HTML dans la variable html_data
html_data = response.text

# (Optionnel) afficher les 500 premiers caractères pour vérifier
print(html_data)

Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


Parse the html data using `beautiful_soup` using parser i.e `html5lib` or `html.parser`.


In [ ]:
from bs4 import BeautifulSoup

# Utiliser html.parser (par défaut inclus avec Python)
soup = BeautifulSoup(html_data, "html.parser")

# OU utiliser html5lib (nécessite l'installation de html5lib)
# soup = BeautifulSoup(html_data, "html5lib")

# Vérification rapide : afficher le titre de la page
print(soup.title.text)

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Step-by-step instructions</summary>

```

Here are the step-by-step instructions:

1. Create an Empty DataFrame
2. Find the Relevant Table
3. Check for the Tesla Quarterly Revenue Table
4. Iterate Through Rows in the Table Body
5. Extract Data from Columns
6. Append Data to the DataFrame

```
</details>


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1

We are focusing on quarterly revenue in the lab.
```

</details>


In [ ]:
!pip install lxml

In [ ]:
import pandas as pd
from bs4 import BeautifulSoup

# 1. Parser le HTML
soup = BeautifulSoup(html_data, "html.parser")

# 2. Créer le DataFrame vide
tesla_revenue = pd.DataFrame(columns=["Date", "Revenue"])

# 3. Identifier le bon tableau
tables = soup.find_all("table")
for table in tables:
    if "Tesla Quarterly Revenue" in table.text:
        target_table = table
        break

# 4. Itérer sur les lignes du tableau
rows = target_table.find_all("tr")[1:]  # On saute l'en-tête

for row in rows:
    cols = row.find_all("td")
    if len(cols) == 2:
        date = cols[0].text.strip()
        revenue = cols[1].text.strip()

        # Exclure les valeurs manquantes ou "–"
        if date != "" and revenue != "" and revenue != "–":
            revenue_clean = revenue.replace("$", "").replace(",", "")
            if revenue_clean.isnumeric():  # Évite les valeurs comme "N/A"
                tesla_revenue = pd.concat([
                    tesla_revenue,
                    pd.DataFrame([{"Date": date, "Revenue": revenue_clean}])
                ], ignore_index=True)

# 5. Réinitialiser les index
tesla_revenue.reset_index(drop=True, inplace=True)

# 6. Afficher les 5 dernières lignes
print(tesla_revenue.tail())

Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [ ]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"",regex=True)

Execute the following lines to remove an null or empty strings in the Revenue column.


In [ ]:
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [ ]:
# Afficher les 5 dernières lignes
tesla_revenue.tail()

## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [ ]:
import yfinance as yf

# Créer l'objet Ticker pour GameStop
gme = yf.Ticker("GME")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to ` "max" ` so we get information for the maximum amount of time.


In [ ]:
# Extraire l'historique complet des cours de GameStop
gme_data = gme.history(period="max")

# Réinitialiser l'index pour que la colonne Date devienne normale
gme_data.reset_index(inplace=True)

# Afficher les premières lignes pour vérifier
gme_data.head()

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [ ]:
import yfinance as yf
import pandas as pd

# Créer l'objet Ticker pour GameStop
gme = yf.Ticker("GME")

# Extraire l'historique complet des cours
gme_data = gme.history(period="max")

# Réinitialiser l'index pour que la colonne Date devienne normale
gme_data.reset_index(inplace=True)

# Afficher les 5 premières lignes
gme_data.head()

## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data_2`.


In [ ]:
import requests

# URL de la page à télécharger
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"

# Envoyer la requête GET
response = requests.get(url)

# Sauvegarder le texte de la réponse dans la variable html_data_2
html_data_2 = response.text

# (Optionnel) Afficher les 500 premiers caractères pour vérifier
print(html_data_2[:500])


In [ ]:
################################# Étape 0: Importer les bibliothèques nécessaires
import requests
import pandas as pd
from bs4 import BeautifulSoup

# Étape 1: Télécharger la page web et stocker dans la variable html_data_2
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"
response = requests.get(url)
html_data_2 = response.text

# Étape 2: Parser le HTML avec BeautifulSoup
soup = BeautifulSoup(html_data_2, "html.parser")

# Étape 3: Créer un DataFrame vide
gme_revenue = pd.DataFrame(columns=["Date", "Revenue"])

# Étape 4: Trouver le tableau avec le titre "GameStop Quarterly Revenue"
tables = soup.find_all("table")

# Identifier le bon tableau en lisant les en-têtes
for table in tables:
    if "GameStop Quarterly Revenue" in table.text:
        target_table = table
        break

# Étape 5: Parcourir les lignes du corps du tableau
for row in target_table.find_all("tr")[1:]:  # Ignorer l'en-tête
    cols = row.find_all("td")
    if len(cols) == 2:
        date = cols[0].text.strip()
        revenue = cols[1].text.strip().replace("$", "").replace(",", "")
        if revenue != "":
            # Étape 6: Ajouter la ligne au DataFrame
            gme_revenue = pd.concat([
                gme_revenue,
                pd.DataFrame([[date, float(revenue)]], columns=["Date", "Revenue"])
            ], ignore_index=True)

# Afficher les 5 dernières lignes pour vérification
print(gme_revenue.tail())

Parse the html data using `beautiful_soup` using parser i.e `html5lib` or `html.parser`.


In [ ]:
!pip install html5lib

In [ ]:
from bs4 import BeautifulSoup

# Parser les données HTML avec html.parser
soup_2 = BeautifulSoup(html_data_2, "html.parser")

# Vérifier le titre de la page pour s'assurer que le parsing a fonctionné
print(soup_2.title.text)

In [ ]:
# Trouver le tableau (supposons qu'il n'y a qu'un seul tableau dans la page)
table = soup.find("table")

# Extraire toutes les lignes
rows = table.find_all("tr")

# Extraire les données dans une liste de listes
gme_revenue_data = []
for row in rows:
    cols = row.find_all("td")
    cols = [col.text.strip() for col in cols]  # enlever les espaces inutiles
    if cols:
        gme_revenue_data.append(cols)

# Afficher les données extraites
for data in gme_revenue_data:
    print(data)

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column.


> **Note: Use the method similar to what you did in question 2.**  


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [ ]:
pip install lxml

In [ ]:
import pandas as pd
from bs4 import BeautifulSoup

soup = BeautifulSoup(html_data_2, "html.parser")
table = soup.find("table")
rows = table.find_all("tr")

data = []
for row in rows[1:]:  # Ignorer l'en-tête
    cols = row.find_all("td")
    date = cols[0].text.strip()
    revenue = cols[1].text.strip().replace("$", "").replace(",", "")
    data.append([date, revenue])

gme_revenue = pd.DataFrame(data, columns=["Date", "Revenue"])
gme_revenue["Revenue"] = gme_revenue["Revenue"].astype(float)

print(gme_revenue.tail())

Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [ ]:
# Afficher les 5 dernières lignes
gme_revenue.tail()

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. Note the graph will only show data upto June 2021.


<details><summary>Hint</summary>

```

You just need to invoke the make_graph function with the required parameter to print the graphs.The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`.

```
    
</details>


In [ ]:
# Convertir la colonne Date en datetime
tesla_revenue["Date"] = pd.to_datetime(tesla_revenue["Date"])

# Vérifier les types
print(tesla_revenue.dtypes)

In [ ]:
make_graph(tesla_data, tesla_revenue, "Tesla Stock Price and Revenue (up to June 2021)")

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


<details><summary>Hint</summary>

```

You just need to invoke the make_graph function with the required parameter to print the graphs.The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`

```
    
</details>


In [ ]:
import pandas as pd

# Convertir la colonne Date en datetime si ce n'est pas déjà fait
gme_revenue["Date"] = pd.to_datetime(gme_revenue["Date"])
 # Espacer les étiquettes de l'axe X (par pas de 500 points)
    ax1.set_xticks(stock_data['Date'][::500])
    ax1.tick_params(axis='x', rotation=45)
# Utiliser la fonction make_graph pour tracer GameStop
make_graph(gme_data, gme_revenue, "GameStop Stock Price and Revenue (up to June 2021)")

In [ ]:
make_graph(gme_data, gme_revenue, "GameStop Stock Price and Revenue (up to June 2021)")

<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
